spec.bs

```diff
@@ -116,6 +132,49 @@ if (document.wasDiscarded) {
 }
 </pre>
 
+Feature Policies {#feature-policies}
+=====================
+To improve CPU utilization of [=nested browsing contexts=] that are not critical to a page's usability two feature policies are defined:
+ * execution-while-not-rendered, which has a [=default allowlist=] of '*'
+ * execution-while-out-of-viewport, which has a [=default allowlist=] of '*'
+
+The <dfn export>execution-while-not-rendered</dfn> policy controls whether [=task source/freezable=]
+tasks should execute for [=nested browsing contexts=] whose [=browsing context container=] is not [=being rendered=]. ie. `display: none`.
+
+The <dfn export>execution-while-out-of-viewport</dfn> policy controls whether [=task source/freezable=]
+tasks should execute for [=nested browsing contexts=] whose [=browsing context container=] does not intersect the [=viewport=] according
+to [=compute the intersection of a target element and the root=].
+
+Application of these policies is accomplished by placing the document (and its decendants) in a [=/frozen=] state when the following
+conditions have been met:
+ * [=iframe load event steps=] have been run
+ * the policy is disabled for the document
+ * relevant policy condition applies (not rendered or scrolled out of view)
+
+If these conditions aren't met the document will be in the [=/unfrozen=] state.
+
+<pre class="example highlight">
+
+// The iframe will be frozen immediately after it is loaded.
+
+&lt;iframe allow="execution-while-not-rendered 'none'"
+  src="subframe.html" style="display:none"&gt;&lt;/iframe&gt;
+</pre>
+
+To run the <dfn export>update document frozenness steps</dfn> for a {{Document}} <var>document</var>,
+run the following steps:
+1. If <var>document</var>'s [=document/browsing context=] is not a [=nested browsing context=], return.
+1. If <var>document</var>'s [=document/readiness=] is not "complete", return.
+1. Let <var>element</var> be <var>document</var>'s [=nested browsing context=]'s [=browsing context container=].
+1. Let <var>frozenness</var> be unset.
+1. If <var>document</var> is not [=allowed to use=] the [=execution-while-not-rendered=] feature then:
+    1. If <var>element</var>'s is not [=being rendered=],  set <var>frozenness</var>'s to true.
+    1. Otherwise, set <var>frozenness</var> to false.
+1. Otherwise if <var>document</var> is not [=allowed to use=] the [=execution-while-out-of-viewport=] feature then:
+    1. If <var>element</var>'s does not intersect the [=viewport=] according to [=compute the intersection of a target element and the root=], set <var>frozenness</var> to true.
+    1. Otherwise, set <var>frozenness</var> to false.
+1. If <var>frozenness</var> does not equal <var>document</var>'s [=FROZENNESS=] state, [=queue a task=] to be set <var>document</var>'s [=FROZENNESS=] to <var>frozenness</var>.
```

Comment:
```suggestion
1. If <var>frozenness</var> does not equal <var>document</var>'s [=Document/frozenness=] state, [=queue a task=] to set <var>document</var>'s [=Document/frozenness=] to <var>frozenness</var>.
```

Reply:
What task source should this task be queued on? DOM manipulation task source is a reasonable default here, but I'm curious which one you'd be implementing in Chrome?

Reply:
In Chrome we actually send this to the browser process to handle (it is the source of all truth for frozenness) to then post back to the individual frame. So in chrome it occurs back on an IPC thread.

Reply:
On what task source do you send this message?

Reply:
It is the same one the freeze event is sent on when freezing from the browser side (the internal IPC task runner). How does that map to these specs.

Reply:
Perhaps you could link to the code. My understanding is we have a variety of task source constants in use.

If we're not posting a task in Chrome (and posting a task requires a task source), then we're violating the spec, and need to rethink this approach.

Reply:
Here is the code...
This is steps [2,4,5,6](https://cs.chromium.org/chromium/src/content/browser/frame_host/render_frame_host_impl.cc?sq=package:chromium&g=0&l=6787)

Which are in fact happening in the browser process. The queue task is the message back to the renderer, which is then received on a mojo channel bound to the IPC task runner which is then processed [here](https://cs.chromium.org/chromium/src/third_party/blink/renderer/core/frame/local_frame.cc?sq=package:chromium&q=LocalFrame::SetLife&g=0&l=1729)

Reply:
Hmm, that is pretty opaque, indeed. It seems a bit related to the discussion at https://groups.google.com/a/chromium.org/forum/#!topic/platform-architecture-dev/grbXI-bh6Rk , where Haraken says "If the task is used for speced tasks (and probably non-speced tasks), use the speced task types", but it's not obvious where in the code the task types even appear. Maybe @zetafunction, as the other person in that thread, could help us out?

Reply:
Separately, it seems to me that we shouldn't be queuing a task here at all. The iframe load event steps are already run inside a queued task, i.e. "on the main thread", so we should be able to synchronously change the frozenness state here.

Can Chrome's implementation change to do this state change during the same task that fires the load event? It seems better if there's not an opportunity for arbitrary tasks to run between firing the load even and freezing the iframe.

Reply:
So I added the queue a task because it really is an IPC from the browser process to the renderer (which is a queue task itself) and I didn't want the spec to make assumptions about the lifecycle of the states. 

So all of this text misses that these are potentially cross origin frames that are implemented in separate processes. Looking at a browsing context's rendering container element could happen in a different process. I tried my best to write an algorithm that approximates the way it is implemented but it is difficult with the process abstraction not present in the specs.

Reply:
Sure, but can't you fire the load event and update the state in a single task, instead of nested ones?

Reply:
Not really cause this happens on the browser side too. The [loaded](https://cs.chromium.org/chromium/src/content/browser/frame_host/render_frame_host_impl.cc?q=render_frame_host_imp&sq=package:chromium&g=0&l=3444) event is processed on the browser side and the frame frozen state is adjusted and then reflected in the page.

Reply:
So there is no way, in Chrome's current architecture, to avoid arbitrary tasks running between the load event, and the frozennness state change? That's frustrating, and it's something we should try to discuss with other browsers before enshrining that kind of limitation in the spec.

Reply:
Because the policy is based on other attributes controlled by parent frames I don't see how (display and visibility in viewport) I'm not sure this situation could be any better. For example an parent frame could only apply the "display: none" attribute on the load step anyways so apps will need to deal with that anyways.

Reply:
Maybe we're talking past each other. Let me spell it out in more detail.

The spec in this PR says when an iframe gets inserted, omitting some other steps, we:

1. Queue a task on the DOM manipulation task source to:
  1. Fire a load event (*)
  1. Queue a task on the ??? task source to:
      1. change the frozenness to true if it's display: none. (**)

I'm proposing the spec should instead say that when an iframe gets inserted, we:

1. Queue a task on the DOM manipulation task source to:
  1. Fire a load event (*)
  1. Change the frozenness to true if it's display: none. (**)

Consider the situation written in this PR as an example:

```html
<iframe src="foo.html" style="display: none" allow="execution-while-not-rendered 'none'"></iframe>
```

With the spec in this PR, between (*) and (**), arbitrary tasks can run, including author JavaScript code, and including freezeable tasks associated with the iframe's document---the very things we're trying to prevent. This is because of all the other tasks in the event loop get a chance to run before the (**) task's turn comes around.

With the spec as I'm proposing, (*) and (**) happen together in the same task, so execution is actually prevented.

I'm hoping it's possible to write code in Chrome to ensure this latter version.

Reply:
Ok I think I can make this work in terms of the load event. I've removed the post a task.